File: frontend/src/api/crm.ts
/**
 * Stratum AI - CRM Integration API
 *
 * Talks to the tenant-scoped CRM surface at /integrations/crm. The tenant is
 * taken from the caller's token, so nothing here sends a tenant id.
 *
 * Field names are the API's, unchanged. The client sends no case transform
 * (see api/client.ts), so renaming them here would mean maintaining a second
 * spelling of every field for no gain.
 *
 * This module previously declared a much wider surface — contact detail,
 * contact journey, deal detail, pipeline metrics, writeback history, writeback
 * retry, connect, disconnect, connection settings — against endpoints that
 * were never built, and typed it with fields the backend does not store
 * (contact email and name are held only as SHA256 hashes). Nothing rendered
 * any of it. Those declarations are gone rather than left pointing at 404s;
 * add them back alongside the endpoint when a view needs one.
 */

import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
import { apiClient, ApiResponse, PaginatedResponse } from './client'

// =============================================================================
// Types
// =============================================================================

export type CRMProvider = 'hubspot' | 'pipedrive'
export type CRMConnectionStatus = 'pending' | 'connected' | 'disconnected' | 'error' | 'syncing'
export type DealStage = 'lead' | 'qualified' | 'proposal' | 'negotiation' | 'won' | 'lost'

export interface CRMConnection {
  id: string
  provider: CRMProvider
  status: CRMConnectionStatus
  status_message?: string | null
  provider_account_id?: string | null
  provider_account_name?: string | null
  sync_contacts: boolean
  sync_deals: boolean
  sync_companies: boolean
  last_sync_at?: string | null
  last_sync_status?: string | null
  last_sync_contacts_count?: number | null
  last_sync_deals_count?: number | null
  created_at: string
  updated_at: string
}

/**
 * A synced contact. Carries no email, name, phone or company: crm_contacts
 * stores only email_hash and phone_hash, so there is no plaintext to return.
 */
export interface CRMContact {
  id: string
  crm_contact_id: string
  lifecycle_stage?: string | null
  lead_source?: string | null
  first_touch_campaign_id?: string | null
  last_touch_campaign_id?: string | null
  first_touch_ts?: string | null
  last_touch_ts?: string | null
  touch_count?: number | null
  stratum_quality_score?: number | null
}

/** A synced deal. Value is in cents; the float `amount` column is deprecated. */
export interface CRMDeal {
  id: string
  crm_deal_id: string
  deal_name?: string | null
  stage?: string | null
  stage_normalized?: DealStage | null
  amount_cents?: number | null
  currency: string
  close_date?: string | null
  expected_close_date?: string | null
  is_won: boolean
  is_closed: boolean
  won_at?: string | null
  attributed_campaign_id?: string | null
  attributed_platform?: string | null
}

export interface PipelineSummary {
  /** 'not_connected' when the tenant has no CRM connection at all. */
  status: string
  stage_counts: Record<string, number>
  stage_values: Record<string, number>
  total_pipeline_value: number
  total_won_value: number
  won_deal_count: number
  last_sync_at?: string | null
}

export interface WritebackConfig {
  id: string
  connection_id: string
  enabled: boolean
  sync_contacts: boolean
  sync_deals: boolean
  auto_sync_enabled: boolean
  sync_interval_hours: number
  sync_attribution: boolean
  sync_profit_metrics: boolean
  sync_touchpoint_count: boolean
  properties_created: boolean
  last_sync_at?: string | null
  last_sync_status?: string | null
  last_sync_contacts?: number | null
  last_sync_deals?: number | null
  last_sync_errors?: number | null
  next_sync_at?: string | null
}

export interface WritebackConfigUpdate {
  enabled?: boolean
  sync_contacts?: boolean
  sync_deals?: boolean
  auto_sync_enabled?: boolean
  sync_interval_hours?: number
  sync_attribution?: boolean
  sync_profit_metrics?: boolean
  sync_touchpoint_count?: boolean
}

export interface SyncTriggerResult {
  connection_id: string
  provider: CRMProvider
  /** 'queued' on success; 'not_connected' or 'unsupported' otherwise. */
  status: string
  contacts_synced: number
  deals_synced: number
  message?: string | null
}

// =============================================================================
// API Functions
// =============================================================================

export const crmApi = {
  getConnections: async () => {
    const response = await apiClient.get<ApiResponse<CRMConnection[]>>(
      '/integrations/crm/connections'
    )
    return response.data.data
  },

  triggerSync: async (connectionId: string) => {
    const response = await apiClient.post<ApiResponse<SyncTriggerResult>>(
      `/integrations/crm/connections/${connectionId}/sync`
    )
    return response.data.data
  },

  getContacts: async (params?: { limit?: number; offset?: number; lifecycle_stage?: string }) => {
    const response = await apiClient.get<ApiResponse<PaginatedResponse<CRMContact>>>(
      '/integrations/crm/contacts',
      { params }
    )
    return response.data.data
  },

  getDeals: async (params?: { limit?: number; offset?: number; stage?: DealStage }) => {
    const response = await apiClient.get<ApiResponse<PaginatedResponse<CRMDeal>>>(
      '/integrations/crm/deals',
      { params }
    )
    return response.data.data
  },

  getPipelineSummary: async () => {
    const response = await apiClient.get<ApiResponse<PipelineSummary>>(
      '/integrations/crm/pipeline/summary'
    )
    return response.data.data
  },

  getWritebackConfig: async () => {
    const response = await apiClient.get<ApiResponse<WritebackConfig | null>>(
      '/integrations/crm/writeback/config'
    )
    return response.data.data
  },

  updateWritebackConfig: async (config: WritebackConfigUpdate) => {
    const response = await apiClient.put<ApiResponse<WritebackConfig>>(
      '/integrations/crm/writeback/config',
      config
    )
    return response.data.data
  },
}

// =============================================================================
// React Query Hooks
// =============================================================================

export function useCRMConnections() {
  return useQuery({
    queryKey: ['crm', 'connections'],
    queryFn: crmApi.getConnections,
  })
}

export function useTriggerCRMSync() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: crmApi.triggerSync,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['crm'] })
    },
  })
}

export function useCRMContacts(params?: {
  limit?: number
  offset?: number
  lifecycle_stage?: string
}) {
  return useQuery({
    queryKey: ['crm', 'contacts', params],
    queryFn: () => crmApi.getContacts(params),
  })
}

export function useCRMDeals(params?: { limit?: number; offset?: number; stage?: DealStage }) {
  return useQuery({
    queryKey: ['crm', 'deals', params],
    queryFn: () => crmApi.getDeals(params),
  })
}

export function usePipelineSummary() {
  return useQuery({
    queryKey: ['crm', 'pipeline', 'summary'],
    queryFn: crmApi.getPipelineSummary,
  })
}

export function useWritebackConfig() {
  return useQuery({
    queryKey: ['crm', 'writeback', 'config'],
    queryFn: crmApi.getWritebackConfig,
  })
}

export function useUpdateWritebackConfig() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: crmApi.updateWritebackConfig,
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['crm', 'writeback'] })
    },
  })
}

File: backend/app/schemas/crm.py
# =============================================================================
# Stratum AI - Tenant-facing CRM Schemas
# =============================================================================
"""Schemas for the tenant-scoped CRM surface under ``/integrations/crm``.

Distinct from the response models inside ``api/v1/endpoints/integrations.py``,
which serve the superadmin surface: those take a ``tenant_id`` query parameter
and are guarded by ``require_super_admin``, while everything here derives the
tenant from the caller's own token.

Two deliberate omissions. The second is data that does not exist; the first is
data this surface must not carry, which is not the same thing:

* ``CRMContactRead`` exposes no email, name, phone or company. ``CRMContact``
  has no plaintext column for any of them — only ``email_hash`` and
  ``phone_hash``, SHA256 digests kept for identity matching — and the hashes are
  not useful to a UI, so neither is serialised.

  This is not the same as saying the plaintext is absent from the database.
  ``CRMContact.raw_properties`` holds HubSpot's whole property payload, which
  ``hubspot_sync`` requests with ``email``, ``phone``, ``firstname`` and
  ``lastname`` in the list. Nothing serialises it and this schema must keep it
  that way; see docs/05-operations/crm-contact-pii-decision.md, which is open.
* Deal money is ``amount_cents`` only. ``CRMDeal.amount`` is marked DEPRECATED
  in the model ("should not be used for calculations"), so exposing it would
  invite a float rounding bug in the client.
"""

from datetime import date, datetime
from typing import Dict, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class CRMConnectionRead(BaseModel):
    """A tenant's CRM connection, without any token material."""

    id: UUID
    provider: str
    status: str
    status_message: Optional[str] = None
    provider_account_id: Optional[str] = None
    provider_account_name: Optional[str] = None

    sync_contacts: bool
    sync_deals: bool
    sync_companies: bool

    last_sync_at: Optional[datetime] = None
    last_sync_status: Optional[str] = None
    last_sync_contacts_count: Optional[int] = None
    last_sync_deals_count: Optional[int] = None

    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CRMContactRead(BaseModel):
    """A synced CRM contact. Identity fields are hashed and never returned."""

    id: UUID
    crm_contact_id: str
    lifecycle_stage: Optional[str] = None
    lead_source: Optional[str] = None

    first_touch_campaign_id: Optional[str] = None
    last_touch_campaign_id: Optional[str] = None
    first_touch_ts: Optional[datetime] = None
    last_touch_ts: Optional[datetime] = None
    touch_count: Optional[int] = None

    stratum_quality_score: Optional[float] = None

    model_config = ConfigDict(from_attributes=True)


class CRMDealRead(BaseModel):
    """A synced CRM deal with its attribution, valued in cents."""

    id: UUID
    crm_deal_id: str
    deal_name: Optional[str] = None
    stage: Optional[str] = None
    stage_normalized: Optional[str] = None

    amount_cents: Optional[int] = None
    currency: str

    close_date: Optional[date] = None
    expected_close_date: Optional[date] = None
    is_won: bool
    is_closed: bool
    won_at: Optional[datetime] = None

    attributed_campaign_id: Optional[str] = None
    attributed_platform: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class PipelineSummaryRead(BaseModel):
    """Deal counts and values by stage for the caller's tenant."""

    status: str
    stage_counts: Dict[str, int] = Field(default_factory=dict)
    stage_values: Dict[str, float] = Field(default_factory=dict)
    total_pipeline_value: float = 0
    total_won_value: float = 0
    won_deal_count: int = 0
    last_sync_at: Optional[str] = None


class WritebackConfigRead(BaseModel):
    """Writeback settings and the outcome of the last run."""

    id: UUID
    connection_id: UUID

    enabled: bool
    sync_contacts: bool
    sync_deals: bool
    auto_sync_enabled: bool
    sync_interval_hours: int

    sync_attribution: bool
    sync_profit_metrics: bool
    sync_touchpoint_count: bool

    properties_created: bool
    last_sync_at: Optional[datetime] = None
    last_sync_status: Optional[str] = None
    last_sync_contacts: Optional[int] = None
    last_sync_deals: Optional[int] = None
    last_sync_errors: Optional[int] = None
    next_sync_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class WritebackConfigUpdate(BaseModel):
    """Editable writeback settings.

    Every field is optional so a caller can PATCH-style update one toggle.
    Carries no tenant_id or connection_id: both are resolved from the
    authenticated tenant, never from the request body.
    """

    enabled: Optional[bool] = None
    sync_contacts: Optional[bool] = None
    sync_deals: Optional[bool] = None
    auto_sync_enabled: Optional[bool] = None
    # Bounded because the value feeds next_sync_at arithmetic in
    # run_scheduled_writebacks; 0 would re-dispatch on every sweep.
    sync_interval_hours: Optional[int] = Field(default=None, ge=1, le=720)
    sync_attribution: Optional[bool] = None
    sync_profit_metrics: Optional[bool] = None
    sync_touchpoint_count: Optional[bool] = None


class SyncTriggerRead(BaseModel):
    """Result of asking for a sync on one connection."""

    connection_id: UUID
    provider: str
    status: str
    contacts_synced: int = 0
    deals_synced: int = 0
    message: Optional[str] = None
